resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala

```diff
@@ -267,30 +271,56 @@ private[spark] object KubernetesUtils extends Logging {
    }
   }
 
-  def uploadFileUri(uri: String, conf: Option[SparkConf] = None): String = {
-    conf match {
-      case Some(sConf) =>
-        if (sConf.get(KUBERNETES_FILE_UPLOAD_PATH).isDefined) {
-          val fileUri = Utils.resolveURI(uri)
-          try {
-            val hadoopConf = SparkHadoopUtil.get.newConfiguration(sConf)
-            val uploadPath = sConf.get(KUBERNETES_FILE_UPLOAD_PATH).get
-            val fs = getHadoopFileSystem(Utils.resolveURI(uploadPath), hadoopConf)
-            val randomDirName = s"spark-upload-${UUID.randomUUID()}"
-            fs.mkdirs(new Path(s"${uploadPath}/${randomDirName}"))
-            val targetUri = s"${uploadPath}/${randomDirName}/${fileUri.getPath.split("/").last}"
-            log.info(s"Uploading file: ${fileUri.getPath} to dest: $targetUri...")
-            uploadFileToHadoopCompatibleFS(new Path(fileUri.getPath), new Path(targetUri), fs)
-            targetUri
-          } catch {
-            case e: Exception =>
-              throw new SparkException(s"Uploading file ${fileUri.getPath} failed...", e)
+  def getUploadPath(conf: SparkConf, client: KubernetesClient): (FileSystem, String) = {
+    conf.get(KUBERNETES_FILE_UPLOAD_PATH) match {
+      case Some(path) =>
+        val hadoopConf = new Configuration()
+        // When spark.kubernetes.file.upload.path is set, we need a cluster specific hadoop config,
+        // and if we use spark.kubernetes.hadoop.configMapName to configure not HADOOP_CONF_DIR, we
+        // should download the configmap to our client side.
+        // 1. add configurations from k8s configmap to hadoopConf
+        conf.get(KUBERNETES_HADOOP_CONF_CONFIG_MAP).foreach { cm =>
+          val hadoopConfFiles = client.configMaps().withName(cm).get().getData.asScala
```

Comment:
You can easily mount a configmap with the pod template feature no need to fetch the config map here.

Reply:
1. `spark.kubernetes.hadoop.configMapName` has no `driver` word in its name, so I guess it is ok to be used all over the spark application including the client process.
2. Using `spark.kubernetes.hadoop.configMapName` and `HADOOP_CONF_DIR` is an **either-or** thing, [check here](https://github.com/apache/spark/blob/5cf2602ccbcada92f11ac715872061f8307d9d70/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStep.scala#L41).  It is more like that they both contain the same thing and are used by client to define the driver pod, so they should be equal. For now, the only difference between them is that the `HADOOP_CONF_DIR`  is in the classpath of client process and the other not.

Reply:
I see that their either-or, but I think if you specify them both it should work.Have you tried that with HADOOP_CONF_DIR? Initially config map was not meant for uploading files from the client machine so it may need to be extended, but I do find fetching the configmap from the cluster redundant.

Reply:
@ifilonenko thoughts here? Do you think we can get the kerberos integration in here?

Reply:
If they are both specified, the main corner case seems like the scenario where they aren't consistent, but this could be checked for.

Reply:
I agree with setting them both, which is my original idea too.

Reply:
@skonto ill make sure to get that PR resolved so that we can use it for testing, yeah